ext/rotoscope/rotoscope.c

```diff
@@ -138,36 +140,114 @@ static rs_tracepoint_t extract_full_tracevals(rb_trace_arg_t *trace_arg,
 
 static bool in_fork(Rotoscope *config) { return config->pid != getpid(); }
 
-static bool tracecmp(rs_tracepoint_t *a, rs_tracepoint_t *b) {
-  return (!rb_str_cmp(a->method_name, b->method_name) &&
-          !rb_str_cmp(a->entity, b->entity) &&
-          a->method_level == b->method_level);
+// Knowingly produces false positives, but not false negatives
+static bool fast_tracecmp(VALUE tpval1, VALUE tpval2) {
+  rb_trace_arg_t *t1 = rb_tracearg_from_tracepoint(tpval1);
+  rb_trace_arg_t *t2 = rb_tracearg_from_tracepoint(tpval2);
+  return rb_tracearg_method_id(t1) == rb_tracearg_method_id(t2);
+}
+
+static bool tracecmp(rs_tracepoint_t *new_call, rs_tracepoint_t *old_call) {
+  if (!rb_str_cmp(new_call->method_name, old_call->method_name) &&
+      !rb_str_cmp(new_call->entity, old_call->entity) &&
+      new_call->method_level == old_call->method_level) {
+    return true;
+  }
+
+  // Some module methods have their call events resolve to `Module`, while their
+  // return events resolve to the proper class invoked. This loosens the
+  // comparison test to make sure we pop the stack correctly in these cases.
+  if (!rb_str_cmp(new_call->method_name, old_call->method_name)) {
+    char *method_name_str = StringValueCStr(old_call->method_name);
+    if (!strcmp(StringValueCStr(old_call->entity), "Module")) {
+      if (!strcmp(method_name_str, "extend") ||
+          !strcmp(method_name_str, "extend_object") ||
+          !strcmp(method_name_str, "initialize")) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+static VALUE unsafe_obj_method(VALUE argv) {
```

Comment:
this is just a one-arg version of `rb_obj_method` so I can use `rb_rescue2` to rescue from `NameError`s.